app/test_plugins/src/main/java/edu/wpi/first/shuffleboard/testplugins/DependentPlugin.java

```diff
@@ -0,0 +1,15 @@
+package edu.wpi.first.shuffleboard.testplugins;
+
+import edu.wpi.first.shuffleboard.api.plugin.Plugin;
+
+/**
+ * A plugin that depends on another plugin.
+ */
+public class DependentPlugin extends Plugin {
+
+  public DependentPlugin() {
+    super("edu.wpi.first.shuffleboard", "Dependent Plugin", "1.0.0", "A plugin that depends on another plugin");
+    addDependency("edu.wpi.first.shuffleboard", "Basic Plugin", "1.0.0");
```

Comment:
If the plugin uses `edu.wpi.first.shuffleboard.testplugins.BasicPlugin.someAPInotInOneDotZeroDotZero()` in this constructor, won't this simply blow up with a `Method Def Not Found` exception making this whole dependency version stuff moot?
Wouldn't it be better to load this information from some resource file that the plugin stores in the JAR? That way you're not going to get classloader issues.

Reply:
If the plugin depends on an API present in version `1.1.0` but not in `1.0.0`, the dependent version string should reflect that. Otherwise you're saying your plugin depends on version A when it _actually_ depends on version B.

Reply:
No, what I'm saying is that, if you have the wrong version on the classpath you're going to get exceptions when the classloader tries to load invalid classes or call methods that don't exist.

Reply:
I think that would only crop up when the plugin class itself has static dependencies or dependencies in the constructor/instantiation blocks/fields.

Like this would crash if version 2 of the dependent plugin isn't loaded, because it'll try to access those classes before the plugin loader can check to see if the plugin can actually be loaded.

```java
public class BadPlugin extends Plugin {

  // crashes if version 2 of the dependent plugin hasn't already been loaded!
  private final Map<DataType, ComponentType> defaultComponents =
    ImmutableMap.of(
        SomeDataTypeFromVersion2, ...
    );

  public BadPlugin() {
    ...
    addDependency("...", "...", "2.0.0");
  }

  @Override
  public Map<DataType, ComponentType> getDefaultComponents() {
    return defaultComponents;
  }
}
```

---

I think there are two solutions here:

1. Only use stuff from other plugins in widgets, data types, or in the plugin _methods_; anywhere but fields or constructors in the plugin class. This is what we currently do in the builtin plugins
2. Store the variables in a static inner class that gets referenced in the plugin methods. The inner class is only loaded when first referenced by one of the methods, eg:

```java
public class MyPlugin extends Plugin {

  public MyPlugin() {
    ...
    addDependency("...", "...", "2.0.0");
  }

  public Map<DataType, ComponentType> getDefaultComponents() {
    return Holder.defaultComponents;
  }

  private static final class Holder {
    private static final Map<DataType, ComponentType> defaultComponents =
      ImmutableMap.of(
        SomeDataTypeFromVersion2, ...
      );
  }

}
```

Of course, there is always the assumption that the plugin developers remember to keep the dependencies that the plugin exports up-to-date with changes to the actual library dependencies. There's not much we can do there, besides maybe wrapping the `loadClass` calls in a try-catch.

Reply:
> Like this would crash if version 2 of the dependent plugin isn't loaded, because it'll try to access those classes before the plugin loader can check to see if the plugin can actually be loaded.

This is why gradle has the `afterEvaluate` block. Because these sorts of problems arise.

You probably need to expose some equivalent to `afterEvaluate` and tell plugin authors not to use dependent types until they are out of the constructor.

Reply:
I think it'll be okay if we keep it as-is and warn plugin authors about this. There are already a few ways of working around it, and the builtin plugins all follow the first solution, so people copying the structure of those plugins shouldn't have any issues at all

Reply:
Another quick solution is annotation based processing.
Then you'd only have the problem if the plugin developer uses a plugin type in a `static` code block.

Reply:
How so? Can you give an example?

Reply:
```java
@Dependencies(
    values = [
        @Dependency("...", "...", "2.0.0"),
        @Dependency("...", "...", "3.0.0"),
    ]
)
class SomePluginName {
}
```

Reply:
Hm. That could work.